chasm/lib/activity/handler.go

```diff
@@ -51,3 +52,191 @@ func (h *handler) StartActivityExecution(ctx context.Context, req *activitypb.St
+				EntityID:    req.GetFrontendRequest().GetRunId(),
+			}),
+			(*Activity).buildPollActivityExecutionResponse,
+			req,
+			nil,
+		)
+	case *workflowservice.PollActivityExecutionRequest_WaitAnyStateChange:
+		return pollActivityExecutionWaitAnyStateChange(ctx, req)
+	case *workflowservice.PollActivityExecutionRequest_WaitCompletion:
+		return pollActivityExecutionWaitCompletion(ctx, req)
+	default:
+		return nil, serviceerror.NewInvalidArgumentf("unexpected wait policy type: %T", req.GetFrontendRequest().GetWaitPolicy())
+	}
+}
+
+// pollActivityExecutionWaitAnyStateChange waits until the activity state has advanced beyond that
+// specified by the submitted token. If no token was submitted, it returns the current state without
+// waiting.
+func pollActivityExecutionWaitAnyStateChange(
+	ctx context.Context,
+	req *activitypb.PollActivityExecutionRequest,
+) (*activitypb.PollActivityExecutionResponse, error) {
+
+	// TODO(dan): do we want to guarantee that response data will differ from that received when the
+	// token was obtained? It's potentially confusing for the server to say "there's been a change"
+	// while returning data in which the change is not apparent.
+
+	refBytesFromToken := req.GetFrontendRequest().
+		GetWaitPolicy().(*workflowservice.PollActivityExecutionRequest_WaitAnyStateChange).
+		WaitAnyStateChange.GetLongPollToken()
+
+	var lastSeenRef chasm.ComponentRef
+	if refBytesFromToken != nil {
+		var err error
+		lastSeenRef, err = chasm.DeserializeComponentRef(refBytesFromToken)
+		if err != nil {
+			return nil, serviceerror.NewInvalidArgument("invalid long poll token")
+		}
+		if lastSeenRef.NamespaceID != req.GetNamespaceId() ||
+			lastSeenRef.BusinessID != req.GetFrontendRequest().GetActivityId() ||
+			lastSeenRef.EntityID != req.GetFrontendRequest().GetRunId() {
+			// token is inconsistent with request
+			return nil, serviceerror.NewInvalidArgument("invalid long poll token")
+		}
+	} else {
+		// This ref will compare less than currentRef in the comparison below.
+		lastSeenRef = chasm.NewComponentRef[*Activity](chasm.EntityKey{
+			NamespaceID: req.GetNamespaceId(),
+			BusinessID:  req.GetFrontendRequest().GetActivityId(),
+			EntityID:    req.GetFrontendRequest().GetRunId(),
+		})
+	}
+
+	// PollComponent will return an error if lastSeenRef is not consistent with the entity
+	// transition history on this shard, or if the state on this shard is behind the ref after a
+	// reload.
+	// TODO(dan): retryability of these errors
+	response, newRef, err := chasm.PollComponent(
+		ctx,
+		lastSeenRef,
+		func(
+			a *Activity,
+			ctx chasm.Context,
+			req *activitypb.PollActivityExecutionRequest,
+		) (*activitypb.PollActivityExecutionResponse, bool, error) {
+			// TODO(dan): we're walking the tree to construct a ref when all we want here is the
+			// root/entity VT. Would it make sense for Context to provide access to root node?
+			currentRefBytes, err := ctx.Ref(a)
```

Comment:
not sure what's intention here. The predicateFn should be implemented based on the Activity's state, like for example the execution status of the activity, and see if it's in completed state.

The returned Ref shouldn't be decoded by application logic, I should un-export the DeserializeComponentRef function.